### PR TITLE
com.alibaba.csp.sentinel.node.StatisticNode#curThreadNum

### PR DESCRIPTION
**Pull request by linlinisme**
using the LongAdder rather than AtomicInteger to Provides better performance

### Describe what this PR does / why we need it
LongAdder has better performance than AtomicInteger at high concurrency

### Does this pull request fix one issue?
https://github.com/alibaba/Sentinel/issues/744

### Describe how you did it
Using LongAdder replace AtomicInteger 

### Describe how to verify it
I make some test performance  and Provides in my commit

### CODE REVIEW

**Comment by sczyh30** [comment]
Hi, thanks for contributing. Could you please provide us your benchmark data?

**Comment by linlinisme** [comment]
> Hi, thanks for contributing. Could you please provide us your benchmark data?
测试数据如下，各自独立起20个线程，执行如下任务
atomicInteger任务
```
 public void run() {
            long startTime = System.currentTimeMillis();

            for(int i = 0; i < 100000; i++) {
                if (op == 0) {
                    atomicInteger.incrementAndGet();
                } else if (op == 1) {
                    atomicInteger.decrementAndGet();
                }
            }
            long cost = System.currentTimeMillis() - startTime;
            taskCostMap.put(taskId,cost);
        }
```
LongAdder任务
```
 public void run() {
            long startTime = System.currentTimeMillis();
            for(int i = 0; i < 100000; i++){
                if(op == 0) {
                    statisticNode.increaseThreadNum();
                }else if(op == 1){
                    statisticNode.decreaseThreadNum();
                }
            }
            long cost = System.currentTimeMillis() - startTime;
            taskCostMap.put(taskId,cost);
        }
```
提交任务

```
   for(int i = 0; i < count; i++){
              int op = i % 2;
              bizEs2.submit(new StatisticAtomicIntegerTask(atomicInteger,op,i));
              bizEs1.submit(new StatisticLongAdderTask(statisticNode,op,i));
          }
```

测试结果：
count = 20
[StatisticNodeTest] LongAdder totalCost : 80ms
[StatisticNodeTest] AtomicInteger totalCost : 67ms

count = 40
[StatisticNodeTest] LongAdder totalCost : 167ms
[StatisticNodeTest] AtomicInteger totalCost : 228ms

count = 60
[StatisticNodeTest] LongAdder totalCost : 278ms
[StatisticNodeTest] AtomicInteger totalCost : 372ms

count = 80
[StatisticNodeTest] LongAdder totalCost : 287ms
[StatisticNodeTest] AtomicInteger totalCost : 1052ms

count = 100
[StatisticNodeTest] LongAdder totalCost : 822ms
[StatisticNodeTest] AtomicInteger totalCost : 1740ms

任务数不多的情况下，并发不高的情况下。AtomicInteger要稍微好于LongAdder，但两者差距不大。
随着任务数的增越多，线程并发越高的情况下，LongAdder的性能表现远优于AtomicInteger。

File: sentinel-core/src/test/java/com/alibaba/csp/sentinel/node/StatisticNodeTest.java [hunk 1]
```diff
@@ -198,4 +195,125 @@ private static void logNode(StatisticNode node) {
...
+        public StatisticLongAdderTask(StatisticNode statisticNode,int op,int taskId){
+            this.statisticNode = statisticNode;
+            this.op = op;
+            this.taskId = taskId;
+        }
+
+        @Override
+        public void run() {
+            long startTime = System.currentTimeMillis();
+            for(int i = 0; i < 100000; i++){
```
Comment (sczyh30) [inline sentinel-core/src/test/java/com/alibaba/csp/sentinel/node/StatisticNodeTest.java]: Could you please reformat your code according to Alibaba Java Coding Guideline? You may refer to the P3C plugin: https://github.com/alibaba/p3c/tree/master/p3c-formatter

Comment (linlinisme) [reply sentinel-core/src/test/java/com/alibaba/csp/sentinel/node/StatisticNodeTest.java]: 我安装了插件，扫描了对应代码，提示没有发现可疑代码
image

Comment (linlinisme) [reply sentinel-core/src/test/java/com/alibaba/csp/sentinel/node/StatisticNodeTest.java]: 装了插件校验，重新提交了代码。麻烦看下 @sczyh30

**Comment by sczyh30** [comment]
Can you also give the memory footprint data after replacing with `LongAdder`? As it increases throughput in exchange for memory.

**Comment by linlinisme** [comment]
> Can you also give the memory footprint data after replacing with `LongAdder`? As it increases throughput in exchange for memory.

@sczyh30 
The test is as follows:
Create 6000 (Sentinel allows up to the number of entries created) AtomicInteger and LongAdder instances to compare them in memory usage.
Initial creation:

```
  List<AtomicInteger> atomicIntegerList = new ArrayList<>(6000);
         For(int i = 0; i < 6000; i++){
             atomicIntegerList.add(new AtomicInteger(0));
         }

```
Combine jps (find java program pid)
and jmap -histo pid | grep "AtomicInteger" command to get the following result:

num    bytes   class
6002  96032 java.util.concurrent.atomic.AtomicInteger

Two internal AtomicInteger objects may be built into the virtual machine internal startup, but this does not affect the measurement results. We know that 6000 AtomicInteger objects occupy approximately 96,032 bytes.


In the same way, the initial memory footprint of 6000 LongAdder objects is 192,000 bytes.
```
 List<LongAdder> longAdders = new ArrayList<>(6000);
        for(int i = 0; i < 6000; i++){
            longAdders.add(new LongAdder());
        }
```
 num            bytes      class
 6000         192000  com.alibaba.csp.sentinel.slots.statistic.base.LongAdder

Then the high-long concurrent update of the LongAdder object.
  ```
 Runnable runnable = () -> {
            For(int i = 0; i < 6000; i++) {
                longAdders.get(i).increment();
            }
            For(int j = 0; j < 6000; j++){
                longAdders.get(j).decrement();
            }
        };

        ExecutorService executorService = newFixedThreadPool(100);
        For(;;){
            executorService.execute(runnable);
        }
```
LongAdder still keep the same memory usage。This means that 6000 LongAdder objects only occupy 95,968 bytes, or about 93.7 KB, more than 6000 Atomic Integers.

I also got the same result on the performance test tool jprofiler.

image

File: sentinel-core/src/main/java/com/alibaba/csp/sentinel/node/StatisticNode.java [hunk 2]
```diff
@@ -104,7 +105,7 @@
     /**
      * The counter for thread count.
      */
-    private AtomicInteger curThreadNum = new AtomicInteger(0);
+     private LongAdder curThreadNum = new LongAdder();
```
Comment (sczyh30) [inline sentinel-core/src/main/java/com/alibaba/csp/sentinel/node/StatisticNode.java]: Remove the one redundant space?

Comment (linlinisme) [reply sentinel-core/src/main/java/com/alibaba/csp/sentinel/node/StatisticNode.java]: ooh, it had been removed.....

**Comment by sczyh30** [comment]
Thanks for contributing!

**Pull request merged**